hil/ext/switches/juniper.py

```diff
@@ -0,0 +1,294 @@
+
+# Copyright 2013-2017 Massachusetts Open Cloud Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the
+# License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied.  See the License for the specific language
+# governing permissions and limitations under the License.
+"""A switch driver for the juniper QFX series.
+
+But in theory can work for any Juniper switch that supports NETCONF.
+Driver uses the python library PyEZ made available by Juniper.
+By default it sets NETCONF session over ssh, i.e. make RPC calls over ssh.
+For this to work, make sure that the Juniper switch is configured
+appropriately to recieve NETCONF connections. Its default port is 830.
+"""
+
+import re
+import logging
+import schema
+import json
+
+from jnpr.junos import Device
+from jnpr.junos.exception import ConnectError
+from jnpr.junos.utils.config import Config
+
+
+from hil.model import db, Switch
+from hil.migrations import paths
+from os.path import dirname, join
+from hil.migrations import paths
+from hil.ext.switches.junos.config_tables.ConfigTables import InterfaceConfigTable
+logger = logging.getLogger(__name__)
+paths[__name__] = join(dirname(__file__), 'migrations', 'juniper')
+
+
+class juniper(Switch):
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/juniper'
+
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = db.Column(db.Integer, db.ForeignKey('switch.id'), primary_key=True)
+    hostname = db.Column(db.String, nullable=False)
+    username = db.Column(db.String, nullable=False)
+    password = db.Column(db.String, nullable=False)
+
+    @staticmethod
+    def validate(kwargs):
+        schema.Schema({
+            'username': basestring,
+            'hostname': basestring,
+            'password': basestring,
+        }).validate(kwargs)
+
+    def session(self):
+        """Creates necessary object ready to be used by other functions
+        to open a session and manipulate the switch configuration.
+        """
+        dev = Device(host=hostname, user=username, passwd=password)
+        dev.bind(cfg=Config)
+        return dev
+
+    def _commit_config(self, loaded_session):
+        """ Takes an open session that has loaded configuration change. 
+        Commits it if it is correct, otherwise rollback
+        to recent commited state.
+        `loaded_session`: session object after new configuration is loaded.
+        """
+        jun = loaded_session
```

Comment:
why reassign it to some variable?

Reply:
o as to keep the rest of the code uniform. the session name in each function is jun but depending on the context, it might be a session that is not yet opened; or an open session or an opened session that is loaded with configuration changes.